site/content/en/concepts/architecture/_index.md

```diff
@@ -0,0 +1,156 @@
+- kubernetes: resources are read, modified and written back to the apiserver
+  - resources may be updated using 3-way merge (kubectl apply)
+
+### Schema driven
+
+Type or object specific logic should NOT be built into the tool.
+Static resource modifications (e.g. *set*) should be configured using type or object
+metadata (e.g. schema).
+
+OpenAPI is used for resource schema.  Tools may support their own OpenAPI
+extensions which should co-exist with extensions owned by other tools.
+
+Support for new types should be introduced through new OpenAPI definitions rather than
+changes to the tool itself.
+
+- Static per-type and per-object resource transformations should use OpenAPI to tell the
+  tool how to modify a given object
+  - e.g. where to set `image` for `set image` is defined in OpenAPI rather than hard-coded
+  
+- Configuration for individual objects / resources may define custom OpenAPI definitions for
+  that specific instance
+  - e.g. an nginx Deployment's `image` may be restricted to the regular expression `^nginx:.*$`
+
+#### kpt vs kubernetes:
+
+- kpt: OpenAPI read from multiple sources -- can also be inlined into individual
+  configuration objects as comments
+- kubernetes: OpenAPI read from apiserver
+
+### Resource + Controller architecture
+
+The desired state of the system should be expressed as declarative, static data structures
+rather than as executable or interpreted code (DSLs and templates are code).  
+
+Dynamic logic should be built into programs which read configuration as input, modify it
+(creating, updating or deleting resources) and emit configuration as output.
+
+It should be possible to compose multiple tools together through pipelining their input / output.
+
+#### kpt vs kubernetes:
+
+- kpt:
+  - read / write files
+  - triggered explicitly (kpt invocations)
+- kubernetes:
+  - read / write http requests
+  - triggered implicitly (watches)
+
+## IO
+
+Both kpt inputs and outputs should be recognized by Kubernetes project tools, published as
+Kubernetes examples or published by the Kubernetes apiserver.
+
+Examples:
+
+- `kubectl get -o yaml`
+- `kubectl apply -f -`
+- `github.com/kubernetes/examples`
+
+Much like Kubernetes controllers, kpt should be able to read its previous outputs and modify
+them, rather than generating them from scratch -- e.g. read a directory of
+configuration and write back to the same directory.
+
+**Note:** This requires symmetric inputs and outputs.
+
+Dynamic logic may be written using templates or DSLs -- which would not support the read-write
+workflow -- by merging the newly generated template / DSL output resources with the input
+resources.
+
+### Targets
+
+Unlike the Kubernetes control-plane, which reads and writes from the apiserver, kpt
+reads and writes from arbitrary, sources so long as they provide resource configuration.
+
+- Local files
+- Files stored in git
+- Command stdin + stdout
+- Apiserver endpoints
+
+## Configuration as Data vs Configuration as Code
```

Comment:
I think this section should be near the top

Reply:
done